services/controlplane/features/client.go

```diff
@@ -0,0 +1,147 @@
+	"os"
+	"runtime"
+	"sort"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/rudderlabs/rudder-server/config"
+	"github.com/rudderlabs/rudder-server/services/controlplane/identity"
+)
+
+var (
+	DefaultTimeout = 30 * time.Second
+	MaxRetries     = uint64(3)
+)
+
+type OptFn func(c *Client)
+
+func WithURL(url string) OptFn {
+	return func(c *Client) {
+		c.url = url
+	}
+}
+
+func WithHTTPClient(httpClient *http.Client) OptFn {
+	return func(c *Client) {
+		c.client = httpClient
+	}
+}
+
+func WithTimeout(timeout time.Duration) OptFn {
+	return func(c *Client) {
+		c.client.Timeout = timeout
+	}
+}
+
+type Client struct {
+	client   *http.Client
+	ua       string
+	url      string
+	identity identity.Identifier
+}
+
+type payload struct {
+	Components []component `json:"components"`
+}
+
+type component struct {
+	Name     string   `json:"name"`
+	Features []string `json:"features"`
+}
+
+func hostname() string {
+	hostname, err := os.Hostname()
+	if err != nil {
+		hostname = "unknown"
+	}
+	return hostname
+}
+
+func New(identity identity.Identifier, fns ...OptFn) *Client {
+	c := &Client{
+		client: &http.Client{
+			Timeout: DefaultTimeout,
+		},
+		url:      config.GetEnv("CONFIG_BACKEND_URL", "https://api.rudderlabs.com"),
+		identity: identity,
+		ua:       fmt.Sprintf("Go-http-client/1.1; %s; control-plane/features; %s", runtime.Version(), hostname()),
+	}
+
+	for _, fn := range fns {
+		fn(c)
+	}
+
+	return c
+}
+
+type PerComponent map[string][]string
+
+func (c *Client) Send(ctx context.Context, components PerComponent) error {
+	var url string
+
+	switch t := c.identity.(type) {
+	case *identity.Namespace:
+		url = fmt.Sprintf("%s/data-plane/v1/namespaces/%s/settings", c.url, c.identity.ID())
+	case *identity.Workspace:
+		url = fmt.Sprintf("%s/data-plane/v1/workspaces/%s/settings", c.url, c.identity.ID())
+	default:
+		return fmt.Errorf("identity not supported %T", t)
+	}
+
+	payload := payload{
+		Components: []component{},
+	}
+
+	for name, features := range components {
+		payload.Components = append(payload.Components, component{
+			Name:     name,
+			Features: features,
+		})
+	}
+
+	// sort by name, so that the order of components is always the same in every call
+	sort.Slice(payload.Components, func(i, j int) bool { return payload.Components[i].Name < payload.Components[j].Name })
+
+	body, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	backoffWithMaxRetry := backoff.WithContext(backoff.WithMaxRetries(backoff.NewExponentialBackOff(), MaxRetries), ctx)
+	return backoff.Retry(func() error {
+		req, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
+		if err != nil {
+			return err
+		}
+
+		req.Header.Set("Content-Type", "application/json")
+		req.Header.Set("User-Agent", c.ua)
+
+		req.SetBasicAuth(c.identity.BasicAuth())
+
+		resp, err := c.client.Do(req)
+		if err != nil {
+			return err
+		}
+		defer resp.Body.Close()
+
+		b, err := io.ReadAll(resp.Body)
```

Comment:
```suggestion
		b, err := io.ReadAll(resp.Body)
		_ = resp.Body.Close()
```

Optional of course but I wanted to share to gather your thoughts given that in this case it seems very easy to avoid deferring.

```go
package features

import (
	"testing"
)

func doNoDefer(t *int) {
	func() {
		*t++
	}()
}

func doDefer(t *int) {
	defer func() {
		*t++
	}()
}

func BenchmarkDefer(b *testing.B) {
	b.Run("yes", func(b *testing.B) {
		t := 0
		for i := 0; i < b.N; i++ {
			doDefer(&t)
		}
	})

	b.Run("no", func(b *testing.B) {
		t := 0
		for i := 0; i < b.N; i++ {
			doNoDefer(&t)
		}
	})
}
```

With go 1.19 it gives me:

```
goos: linux
goarch: amd64
pkg: github.com/rudderlabs/rudder-server/services/controlplane/features
cpu: 12th Gen Intel(R) Core(TM) i9-12900HK
BenchmarkDefer
BenchmarkDefer/yes
BenchmarkDefer/yes-20         	588573512	         2.031 ns/op
BenchmarkDefer/no
BenchmarkDefer/no-20          	1000000000	         0.2673 ns/op
PASS
```

Perhaps I'm not measuring it appropriately but it seems like some overhead is there :thinking: 

(I'm not saying do not use `defer`, just think twice if it's not really needed. If you have multiple `if/else` with returns then paying the overhead makes sense for the sake of keeping the complexity under control, in any case again totally optional).

Reply:
Interesting, I would not expect this to be that big.

However, the performance overhead of using defer in this case should be minor. Indeed, in this case, it can be easily avoided, but I like it as a consistent pattern across the code. I expect a defer close as soon as a successful request has returned.

For me, the readability/predictability ways the performance when IO is involved.

Reply:
Yes I agree with you, having the defer makes it easier to follow if you're coming from a Go background.

It is a micro optimization on a piece of code that is not really giving us problems so like yourself I think that readability should take precedence for the sake of code maintenability.

I wanted to share this anyway because sometimes I avoid using `defer` myself for this reason (when the code is quite simple and doesn't have returns) plus sharing is caring so :smile: